pygame-a-primer/py_tut_with_images.py

```diff
@@ -0,0 +1,211 @@
+    KEYDOWN,
+    QUIT,
+)
+
+
+# Define the Player object extending pygame.sprite.Sprite
+# Instead of a surface, we use an image for a better looking sprite
+class Player(pygame.sprite.Sprite):
+    def __init__(self):
+        super(Player, self).__init__()
+        self.surf = pygame.image.load("jet.png").convert()
+        self.surf.set_colorkey((255, 255, 255), RLEACCEL)
+        self.rect = self.surf.get_rect()
+
+    # Move the sprite based on keypresses
+    def update(self, pressed_keys):
+        if pressed_keys[K_UP]:
+            self.rect.move_ip(0, -5)
+            move_up_sound.play()
+        if pressed_keys[K_DOWN]:
+            self.rect.move_ip(0, 5)
+            move_down_sound.play()
+        if pressed_keys[K_LEFT]:
+            self.rect.move_ip(-5, 0)
+        if pressed_keys[K_RIGHT]:
+            self.rect.move_ip(5, 0)
+
+        # Keep player on the screen
+        if self.rect.left < 0:
+            self.rect.left = 0
+        elif self.rect.right > 800:
+            self.rect.right = 800
+        if self.rect.top <= 0:
+            self.rect.top = 0
+        elif self.rect.bottom >= 600:
+            self.rect.bottom = 600
+
+
+# Define the enemy object extending pygame.sprite.Sprite
+# Instead of a surface, we use an image for a better looking sprite
+class Enemy(pygame.sprite.Sprite):
+    def __init__(self):
+        super(Enemy, self).__init__()
+        self.surf = pygame.image.load("missile.png").convert()
+        self.surf.set_colorkey((255, 255, 255), RLEACCEL)
+        # The starting position is randomly generated, as is the speed
+        self.rect = self.surf.get_rect(
+            center=(random.randint(820, 900), random.randint(0, 600))
+        )
+        self.speed = random.randint(5, 20)
+
+    # Move the enemy based on speed
+    # Remove it when it passes the left edge of the screen
+    def update(self):
+        self.rect.move_ip(-self.speed, 0)
+        if self.rect.right < 0:
+            self.kill()
+
+
+# Define the cloud object extending pygame.sprite.Sprite
+# Use an image for a better looking sprite
+class Cloud(pygame.sprite.Sprite):
+    def __init__(self):
+        super(Cloud, self).__init__()
+        self.surf = pygame.image.load("cloud.png").convert()
+        self.surf.set_colorkey((0, 0, 0), RLEACCEL)
+        # The starting position is randomly generated
+        self.rect = self.surf.get_rect(
+            center=(random.randint(820, 900), random.randint(0, 600))
+        )
+
+    # Move the cloud based on a constant speed
+    # Remove it when it passes the left edge of the screen
+    def update(self):
+        self.rect.move_ip(-5, 0)
+        if self.rect.right < 0:
+            self.kill()
+
+
+# Initialize pygame
+pygame.init()
+
+# Setup for sounds, defaults are good
+pygame.mixer.init()
+
+# Setup the clock for a decent framerate
+clock = pygame.time.Clock()
+
+# Create the screen object
+# Here we pass it a size of 800x600
+screen = pygame.display.set_mode((800, 600))
```

Comment:
What are your thoughts on using constants here instead? Especially since you use these values elsewhere to refer to the width and height of the viewport:

```python
        elif self.rect.right > 800:
            self.rect.right = 800
```

```python
        elif self.rect.bottom >= 600:
            self.rect.bottom = 600
```

```python
        self.rect = self.surf.get_rect(
            center=(random.randint(820, 900), random.randint(0, 600))
        )
```

Reply:
Makes perfect sense - changes under 8c0cfa2, reformatting under 0550d39.
I also made similar changes to the code in the article for completeness. That required some reformatting and typo fixes as well.